File: convert_sorted_array_to_binary_search_tree.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param num, a list of integers
    # @return a tree node
    def sortedArrayToBST(self, num):
        def dfs(start,end):
            if start > end: return None
            mid = (start + end) / 2
            root = TreeNode(num[mid])
            root.left = dfs(start,mid-1)
            root.right = dfs(mid+1,end)
            return root
        
        return dfs(0,len(num)-1)

File: cleanCode50/ConvertSortedArrayToBinarySearchTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode sortedArrayToBST(int[] nums) {
    	return sortedArrayToBSTHelper(nums, 0, nums.length - 1);
    }

    private TreeNode sortedArrayToBSTHelper(int[] nums, int start, int end) {
    	if (start > end) {
    		return null;
    	}
    	int mid = start + (end - start) / 2;
    	TreeNode root = new TreeNode(nums[mid]);
    	TreeNode left = sortedArrayToBSTHelper(nums, start, mid - 1);
    	TreeNode right = sortedArrayToBSTHelper(nums, mid + 1, end);
    	root.left = left;
    	root.right = right;
    	return root;
    }
}